### ISSUE TITLE
The scanning speed is slow

### ISSUE THREAD

**Issue by lygwy** [dynarithmic/twain_library-java#24]
I didn't find any more information about API calls in the java language in twain_library, especially for operations like callbacks, but luckily I found it in Twain_library-Java. However, I found that the scanning speed of TWAIN_library-Java is slower than that of twain_library. It always stops after scanning one piece, sometimes for 2~3 seconds or even longer, which is obviously unacceptable in practical applications. I want to know the reason and hope to find a solution. I guess there is some parameter to control this speed, but I still can't find it, hope the author can give some guidance

**Comment by dynarithmic** [comment]
You should produce a TWAIN log that contains the actual time stamps of the operations.  See the TWAIN logging demo.  

Other than that, I cannot answer your question about the slow scanning speed without knowing exactly the scanner model, what type of acquisition you are using (File, Native, etc.)

**Comment by dynarithmic** [comment]
You should run the DTWDEMO64U.exe program, as that one has many more options than the SimpleAcquire program.  The call stack log is missing from the log you produced.

I do not have any of the scanners you mentioned, so I can't test this myself.  However, I took a look at the twain_library_demo_log, and I see this sequence:

```
[2024-12-04 14:41:52.769] Thread [39756] Input: 
DSM_Entry(pSource=000002168C381328H, pDest=000002168C3C01E4H, DG_CONTROL, DAT_PENDINGXFERS, MSG_ENDXFER, TW_MEMREF=0000008354FFD2C2H) called
Decoded pSource:
{
    Id=1
    Version Number=1.0
    Version Language=13
    Version Country=1
    Version Info=<?>
    ProtocolMajor=2
    ProtocolMinor=5
    SupportedGroups=DG_CONTROL, DG_IMAGE, DG_AUDIO, DF_DSM2, DF_APP2
    Manufacturer=<?>
    Product Family=<?>
    Product Name=<?>
}

Decoded pDest:
{
    Id=2
    Version Number=21.1
    Version Language=13
    Version Country=1
    Version Info=Aug. 28. 2020
    ProtocolMajor=2
    ProtocolMinor=4
    SupportedGroups=DG_CONTROL, DG_IMAGE, DF_DS2
    Manufacturer=Panasonic
    Product Family=Panasonic Document Scanner
    Product Name=Panasonic KV-SL1077/66/56 MK2
}


TW_MEMREF is TW_PENDINGXFERS:
{
    Count=0
    EOJ=0
}
[2024-12-04 14:41:53.623] Thread [39756] Output: return code from DSM=0 (TWRC_SUCCESS)

```
Note the starting time stamp of `[2024-12-04 14:41:52.769]` and the ending time stamp of `[2024-12-04 14:41:53.623]` -- that is close to 1 second.

The operation that is occurring in that time frame is the following processing by your TWAIN driver of the TWAIN triplet `DG_CONTROL / DAT_PENDINGXFERS / MSG_ENDXFER`.  This triplet is querying your device to test if there are more pages to scan.  Your device responds in the `Output` (that is the `TW_PENDINGXFERS` value) that there are no more pages to scan.  

There are no operations in between the call to query your device, and your device responding back with the results.  Thus the conclusion from the log it is the device that is causing the delay, not DTWAIN.  There is a similar issue starting at timestamps `[2024-12-04 14:41:51.911]` and `[2024-12-04 14:41:52.733]`, where it takes close to a second for the device to respond that "there are no more images to scan".  

If you take the log and place into a good text editor, and only showed the lines with "Thread" in them, you will see the order of the timings, and clearly the big differences are 

1) The time it takes to show the UI when invoked (`DG_CONTROL, DAT_USERINTERFACE, MSG_ENABLEDS` at time stamp `[2024-12-04 14:41:37.810]` and `[2024-12-04 14:41:39.041]`).  
2) (My assumption): The time it takes the user to press the "Scan" button, plus the time it took the device to move the paper through the scanner (The transitioning of the `TWRC_NOTDSEVENT` and `TWRC_DSEVENT`).   So if my assumption is correct, it takes about 6 to 8 seconds for these actions to occur.
3) The time it takes for the device to report whether there are more pages to scan (what was mentioned earlier, taking close to 1 second).

Also, in the log you produced for Java, you are missing this important information that did show up in the non-Java log, and that is the interaction that is going on between DTWAIN and the TWAIN DSM (TWAINDSM.DLL) and your device.  You are only listing the calls made to DTWAIN, but not the communication between DTWAIN and TWAIN itself.

With the added layer of the JNI plus your callback code in Java, it is understandable why the Java code would take longer, but the underlying issue seems to be that your scanner drivers are having an issue with reporting the number of pages that are left to scan.  This could be an issue with your version of TWAINDSM.DLL, but I do not know what version you are using with the logs that you have given me.  This is where using DTWDEMO64U would have these details.

**Comment by dynarithmic** [comment]
Possibly your scanners support the following capabilities (you should give me the full capability list -- you can get it from the logs, running the DTWDEMO64 program, or even get it from Java):

`CAP_AUTOSCAN` and `CAP_MAXBATCHBUFFERS`.  

If your device supports these capabilities, then possibly you can set these values and determine if they make a difference.   You can test, get, and retrieve the maximum number of buffers your device supports by calling the [getMaxBatchBuffers()](https://github.com/dynarithmic/twain_library-java/blob/b0616a438e5ca2db937af1ede2ce1223503d9b68/src/com/dynarithmic/twain/highlevel/capabilityinterface/CapabilityInterface.java#L1214) function from the `TwainSource.getCapabilityInterface()` function (see the `GetSetCapabilities.java` demo program). 

If your device does support these capabilities, and the max batch buffers supported is greater than 0 (or 1), you need to call the `setAutoscan()` first, and then the `setMaxBatchBuffers` second.  Of course, if your device does not support these capabilities, then nothing will happen.

Unfortunately, [the current Java code doesn't set these capabilities when acquiring an image](https://github.com/dynarithmic/twain_library-java/blob/b0616a438e5ca2db937af1ede2ce1223503d9b68/src/com/dynarithmic/twain/highlevel/TwainSource.java#L273).  The `OptionsApplyer` doesn't set the autoscan and maxbatchbuffers.  In the next update of the Java library, I will add this to the options that are automatically applied.

For further information on high-speed scanning possibilities, it is discussed [here](https://www.dynarithmic.com/onlinehelp/dtwain/newversion/High-SpeedImageAcquisition.html) in the DTWAIN documentation.

**Comment by dynarithmic** [comment]
I updated the [development branch](https://github.com/dynarithmic/twain_library-java/tree/development) with the AutoScanningOptions.

Your code could look something like this (before the call to acquire() the pages):

```
TwainSource theSource;
//...
// Assume that theSource has already been selected and opened.
//...
CapabilityInterface ci = theSource.getCapabilityInterface();
if ( ci.isMaxBatchBuffersSupported() )
{
     List<Integer> intList = ci.getMaxBatchBuffers(ci.get());
     int num = intList.get(0);  // See notes after this code sample
     AutoScanningOptions opts = theSource.getAcquireCharacteristics().getAutoScanningOptions();
     opts.enableAutoScan(true).setMaxBatchBuffers(num);
}
```
Note that the call to `getMaxBatchBuffers` may return more than one value.  Those values are either going to be a list of valid buffer values, or if you get 5 values, it more than likely is a range (low, high, step value, current, default).  If you get only one value in the list, then that is the only value available.  Set `num` to the appropriate value.

The test is done initially to make sure your device supports this.

**Comment by lygwy** [comment]
1.You suggested me to run DTWDEMO64U.exe to get more information, I found and ran it in `twain_library\binaries\64bit\64bit_FullDemo\DTWDEMO64U.exe`, but I didn't get the effect I wanted. It still has a short pause while scanning, just like in my previous tests. I printed the log of its operation to the file:

(Supplement: Here I use DTWDEMO64U.exe from twain_library 5.5.4)
[dtwdemo_exe_log.log](https://github.com/user-attachments/files/18018617/dtwdemo_exe_log.log)

2.You mentioned that I only listed calls to DTWAIN, but not the communication between DTWAIN and TWAIN itself, so I reset the log level and rerunned the scanner: `logging.setVerbosity(TwainLogger.LoggerVerbosity.MAXIMUM)`

(But I'm not sure if this is what you mean by "communication between DTWAIN and TWAIN itself," so please enlighten me if I'm wrong.)

(Supplement: The logs here are from twain_library-java 1.5 TinyApplication.java)
[twain-java-demo-log.log](https://github.com/user-attachments/files/18018630/twain-java-demo-log.log)

3.You analyzed the contents of twain_library_demo_log yesterday, and you think that there is a problem with the scanner driver reporting the number of remaining pages. I should note that twain_library_demo_log comes from twain_library 5.5.3, and as I mentioned earlier, the scanning process with the twain_library library is very smooth, one scan after another without any break. I think the problem you mentioned should occur when the scanner program is started, because in the actual test, we did find this problem, when the program runs to `api.DTWAIN_AcquireNative`, the user UI provided by the scanner always appears after 1-2 seconds. I can almost confirm that this is the problem of the scanner manufacturer, because I once tested a scanner in China, and when the program runs to `api.DTWAIN_AcquireNative`, the scanner UI always pops up in time.


4.You mentioned the possible version of TWAINDSM.DLL, so I specially downloaded TWAINDSM.DLL from twain_library 5.5.4, and I found that its latest update was last year. I figured my TWAINDSM.DLL would also be up to date (since I only started using twain a few months ago), but I still re-downloaded it. I put it in C:\Windows\System32 and ran DTWDEMO64U.exe and the result was not what I wanted. I put it back in twain_library-java 1.5 (I reset java.library.path), but the results are still no better.



5.You mentioned `CAP_AUTOSCAN `and `CAP_MAXBATCHBUFFERS`, and I have tested them as suggested. . Unfortunately, `ci isMaxBatchBuffersSupported ()` returns false, `ci.GetMaxBatchBuffers (ci.get())` returns a collection of the empty, Then I won't be able to try the next `setMaxBatchBuffers()`

-----------------------------------------

Forgive me for not replying to your comments in time, because I have been using the test for a long time, and this is the first time for me to contact the docking work of the scanner, so I don't know much about many contents.

I am afraid that I have not clearly stated my problem, so I will restate it:

Suppose I scan ten pieces of A4 paper: ABCDEFGHIJ

In a typical case (for example, with twain_library 5.5.3), the scan would work like this:

`A - 1ms - B - 1ms - C - 1ms .....  -J`

In the current case (for example, with twain_library-java 1.5), the scan would work like this:

`A - 100ms - B - 100ms - C -100ms ......  -J`



In other words, the ideal effect should be: the time interval between the "scan end" of the previous paper and the "scan start" of the following paper is very short and almost imperceptible (such as 1 to 5ms).

The current situation is that the time interval between the "scanning end" of the previous paper and the "scanning start" of the latter paper can be clearly perceived (for example, 100~500ms).


To my wonder, the scanning process of twain_library 5.5.3 is smooth, while the scanning process of Twain_library-Java 1.5 has obvious time intervals, and the scanning process of DTWDEMO64U.exe also has obvious time intervals, which makes me confused. Is there anything special about twain_library?

**Comment by lygwy** [comment]
I just rerun DTWDEMO64U.exe to try to find something new out of it, and I did find something. When I selected Acquire Native for a scan, DTWDEMO64U.exe first gave me a Acquire Setting dialog, I checked "Use Document Feeder" and was surprised to see how smooth the scan process was.



I immediately went to twain_library - global search "Document Feeder" in Java 1.5, several results are obtained, but it seems that only AutoAdjustOptions. EnableSenseMedium () seems to be that can be used, So I added the following code to TinyApplication.java:

AutoAdjustOptions autoAdjustOptions = source.getAcquireCharacteristics().getAutoAdjustOptions();

autoAdjustOptions.enableSenseMedium(true);



Then the result was to my great disappointment, he was still not smooth.



Then I went back to DTWDEMO64U.exe and tried to verify that there was some time between scans without "Use Document Feeder" checked, but then something weird happened. Even if I didn't check "Use Document Feeder", The scanner still works smoothly. I tried it again several times and the results were always smooth, even after I restarted the computer...



I couldn't help but wonder if my memory was messed up, or if the results of my test a few hours earlier were the same? But I clearly remember that DTWDEMO64U.exe did not work so smoothly when scanning...



While I was writing this reply, I opened DTWDEMO64U.exe again, I didn't check "Use Document Feeder" this time, and it actually started generating time intervals again. To make sure of this result, I scanned again without checking "Use Document Feeder" and sure enough there was a time gap. Then I checked "Use Document Feeder" to test again, and the next few scans went smoothly as expected.

Now, I'm even more confused...

This is the log that DTWDEMO64U.exe smoothly scanned:

[dtwdemo_smooth_log.log](https://github.com/user-attachments/files/18019482/dtwdemo_smooth_log.log)


(The non-fluent scan log I provided in the last comment)

**Comment by dynarithmic** [comment]
In the twain-java-demo log, there is this:
```
[2024-12-05 11:01:34.013] Thread [26180] Input: 
DSM_Entry(pSource=000002156A0265C8H, pDest=000002156A070BD4H, DG_CONTROL, DAT_PENDINGXFERS, MSG_ENDXFER, TW_MEMREF=000000B7A9BFCCB2H) called
Decoded pSource:
{
    Id=1
    Version Number=1.0
    Version Language=2
    Version Country=1
    Version Info=<?>
    ProtocolMajor=2
    ProtocolMinor=5
    SupportedGroups=DG_CONTROL, DG_IMAGE, DG_AUDIO, DF_DSM2, DF_APP2
    Manufacturer=<?>
    Product Family=<?>
    Product Name=<?>
}

Decoded pDest:
{
    Id=2
    Version Number=21.1
    Version Language=13
    Version Country=1
    Version Info=Aug. 28. 2020
    ProtocolMajor=2
    ProtocolMinor=4
    SupportedGroups=DG_CONTROL, DG_IMAGE, DF_DS2
    Manufacturer=Panasonic
    Product Family=Panasonic Document Scanner
    Product Name=Panasonic KV-SL1077/66/56 MK2
}


TW_MEMREF is TW_PENDINGXFERS:
{
    Count=0
    EOJ=0
}

[2024-12-05 11:01:36.607] Thread [26180] Output: return code from DSM=0 (TWRC_SUCCESS)

```
It is taking more than 2 seconds for the device to report that there are no pages left to scan.  This is basically the same issue that I mentioned in the previous reply.

Note that this is not really directly a Java issue -- the time stamps above are occurring [here in the C++ code](https://github.com/dynarithmic/twain_library_source/blob/da69d540dac2bfc1054eec25d1c25184dfe0dd71/source/cpp/ctltwainmanager.cpp#L2728).  That one line of code is what the section of the log I pointed out is invoking -- it is the basic communication lane between DTWAIN, the TWAIN DSM, and your scanner driver.   DTWAIN is actually waiting for the function to return back the information (scanner driver -> TwainDSM -> DTWAIN).

The only thing I can say at this point is to try and see if you can call DTWAIN_AcquireNative() directly from Java through the API Handle that you can get from the TWAIN Source.  It doesn't matter if everything hasn't been set up, you should be able to start the scanning process, just to test the speed of the scanning process (right now, the actual image data isn't important).

If directly calling [DTWAIN_AcquireNative()](https://github.com/dynarithmic/twain_library-java/blob/b0616a438e5ca2db937af1ede2ce1223503d9b68/src/com/dynarithmic/twain/DTwainJavaAPI.java#L528) makes a difference, then quite possibly that one (or more) of the options that are being applied by the OptionsApplyer() logic is interfering in the process.  If the scanning process is faster with a direct call, then I would slowly comment out the OptionsApplyer() code to see which options may be causing the issue.

**Comment by lygwy** [comment]
This was definitely an exciting discovery, and I did a scan as you suggested by calling api.DTWAIN_AcquireNative(). I ran multiple alternate tests on source.acquire() and api.DTWAIN_AcquireNative() and almost confirmed that, Calling DTWAIN_AcquireNative() directly does make scanning work smoothly.

```
TwainSession session = new TwainSession();

TwainSource source = session.selectSource();
if (source.isOpened()) {

// source.acquire();

    DTwainJavaAPI api = session.getAPIHandle();
    TwainAcquisitionArray acquisitionArray = api.DTWAIN_AcquireNative(source.getSourceHandle(), 2, -1, true, false);

    if (acquisitionArray != null) {
        com.dtwain.test.displaydialogs.ImageDisplayDialog idg = new com.dtwain.test.displaydialogs.ImageDisplayDialog(acquisitionArray.getAcquisitonArray(), DTwainConstants.FileType.BMP, true);
        idg.setVisible(true);
    }
}
```


So I tried to find out which options were interfering with the scan by commenting out some of the options in OptionsApplyer, but the results were confusing because it sometimes worked well and sometimes didn't when commenting out the same options multiple times. I suspect that calling DTWAIN_AcquireNative() may be affecting the results of the next run. If so, does that mean that some configuration files are being modified each time the program is started? Or is it not because of OptionsApplyer? But this is only my personal guess, the specific reason depends on your verification.



Here are the results of my simple test (based on the following tests, I guess that each run of the program has an impact on the next run of the program) :



Call DTWAIN_AcquireNative(): Smooth scan

Call source.acquire() and comment out all options in TwainSource.startApply() : smooth scan

Call source.acquire() and uncomment all options in TwainSource.startApply() : not smooth scanning

Use source.acquire() and comment out all options in TwainSource.startApply() : not smooth scanning

Call source.acquire() and uncomment all options in TwainSource.startApply() : not smooth scanning

**Comment by dynarithmic** [comment]
I think quite possibly the first time you ran with the options uncommented, this set the scanner in a state where the previous options were used.   

You can try to reset all the capabilities by calling [DTWAIN_SetAllCapsToDefault](https://github.com/dynarithmic/twain_library-java/blob/b0616a438e5ca2db937af1ede2ce1223503d9b68/src/com/dynarithmic/twain/DTwainJavaAPI.java#L360) and see if that makes any difference between scans.

I am working on getting the OptionsApplyer to be a little bit more flexible, where you can specify the options to set as well as the priority of setting the options when acquiring the images.  That won't be available until the next version.

**Comment by lygwy** [comment]
I seem to have found the problem, and my preliminary conclusion from multiple rounds of testing is that the following line of code affects the scanning speed:

`OptionsApplyer.apply(this, this.acquireCharacteristics.getPaperHandlingOptions());`



The DTWAIN_SetAllCapsToDefault method you provided does reset everything. I ran it after the startApply() method during the test. Here are the last few times I tested it:



`startApply() `applies all options and calls `DTWAIN_SetAllCapsToDefault`: Smooth scan

`startApply()` comments all options, applies only `PaperHandlingOptions`, does not call `DTWAIN_SetAllCapsToDefault`: scanning is not smooth

`startApply() `applies all options without calling `DTWAIN_SetAllCapsToDefault`: scanning is not smooth



`startApply()` applies all options and calls `DTWAIN_SetAllCapsToDefault`: Smooth scan

`startApply()` applies all options except `PaperHandlingOptions` and does not call `DTWAIN_SetAllCapsToDefault`: smooth scan

`startApply()` comments all options, applies only `PaperHandlingOptions`, does not call `DTWAIN_SetAllCapsToDefault`: scanning is not smooth



Within the scope of my current ability, I have reached the above results, but the specific conclusion needs to be verified by you.

**Comment by dynarithmic** [comment]
1) Can you run the `TwainCapListingDemo.java` program to get a list of all the capabilities for your scanner?  I don't think I got a listing from you with the capabilities.

2) In [OptionsApplyer.apply](https://github.com/dynarithmic/twain_library-java/blob/b0616a438e5ca2db937af1ede2ce1223503d9b68/src/com/dynarithmic/twain/highlevel/OptionsApplyer.java#L608), can you slowly comment out the various `set` lines to see which aspect of the paper handling causes the issue?

Obviously there is one or more paper handling related functions from [PaperHandlingOptions.java](https://github.com/dynarithmic/twain_library-java/blob/master/src/com/dynarithmic/twain/highlevel/acquirecharacteristics/PaperHandlingOptions.java) that is causing the issue.  I am curious as to which one it is, but need the capability listing to verify.

**Comment by lygwy** [comment]
I ran the `TwainCapListingDemo.java `program and got a list of features:
[TwainCapListingDemo_result.txt](https://github.com/user-attachments/files/18034243/TwainCapListingDemo_result.txt)


In `PaperHandlingOptions' OptionsApplyer.apply()`, after several tests, I found that the problem was` isAutoFeedEnabled()`. The scanner worked smoothly when I made the following changes to the program:

```
boolList.set(0, true);

ci.setAutoFeed(boolList, op);
```

And when I reset it to false, the scanner was having problems again.


I tried to set `AutoFeed `in `TinyApplication.java`, but it didn't seem to work:


```
//....
TwainSource source = session.selectSource();
if (source.isOpened()) {
     CapabilityInterface ci = source.getCapabilityInterface();
     ci.setAutoFeed(Arrays.asList(true), ci.set());
     source.acquire();
}
```

**Comment by dynarithmic** [comment]
Thank you for this information.

In the TinyApplication, you should set the auto feed by using the AcquireCharacteristics:

```
TwainSource source = session.selectSource();
if (source.isOpened()) {
          AcquireCharacteristics ac = source.getAcquireCharacteristics();
          ac.getPaperHandlingOptions().enableAutoFeed(true);
          source.acquire();
}

```
The reason why you should do this using the acquire characteristics is that when `acquire()` is called, the `apply()` logic in `TwainSource` uses the acquire characteristics to set the various capabilities.  So basically your code had physically set the capability in TinyApplication, but it was overwritten as soon as `apply()` was called for `PaperHandlingOptions` when the `acquire()` was done.

In general (as of this version of the library), you should use the AcquireCharacteristics to logically set the capabilities, and let the `apply()` logic actually set the capabilities of the device using the information in the AcquireCharacteristics.  However, you can always use the `CapabilityInterface` to *get* the capability values.

In a future version, this would be more customized, where you can actually set the capabilities (similar to what you did with the TinyApplication) and have the settings "stick" without being overwritten.

**Comment by dynarithmic** [comment]
Can you try the [development version](https://github.com/dynarithmic/twain_library-java/tree/development) without any coding changes on your end (except that the jar file is now 1.5)?  

The scanner should run correctly, but need to verify.  In the meantime for version 1.4 (current version), try setting the `enableAutoFeed(true)`, as mentioned in the previous comment.

**Comment by dynarithmic** [comment]
Also, since your device supports `ICAP_EXTIMAGEINFO`, if you are not interested in the extended image information (barcodes, patchcodes, etc.), then you can turn this off (this may speed up the scanning).

```
TwainSource source;
//...
source.getAcquireCharacteristics().getImageInformationOptions().setExtendedImageInfoAvailable(false);
```
This will turn off the scanner attempting to get information that you may not care about.  

For more information, see the `ExtendedImageInfoDemo.java` demo program.  If anything in the demo is something you are interested in, then call `setExtendedImageInfoAvailable` to `true` (or leave it unset, as the default is `true`).

**Comment by lygwy** [comment]
I have tried the method you provided in TinyApplication and my project respectively(`ac.getPaperHandlingOptions().enableAutoFeed(true)`), and it has proved to be an effective method. I saw the code you submitted for the development branch, but I don't think there seems to be a need to change the version now, as the method you provided is sufficient to solve the problem. In the future, if there are other problems, I may consider looking for a solution from the latest version.

Thank you very much for your help to solve my problem, I think this problem can be closed.
